File: 프로그래머스/lv3/42628. 이중우선순위큐/이중우선순위큐.py
from heapq import heappop, heappush

def solution(operations):
    answer = []
    for x in operations:
        order, num = x.split(' ')
        
        if order == 'I':
            heappush(answer, int(num))
        elif num == '1' and len(answer)>0:
            answer.remove(max(answer))
        elif num == '-1' and len(answer)>0:
            heappop(answer)

    if len(answer)<2:
        return [0,0]
    else:
        return [max(answer), min(answer)]

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;
class Solution {
    public int[] solution(String[] operations) {
        int[] answer = new int[2];
        TreeMap<Integer, Integer> q = new TreeMap<>();
        
        for (String e : operations) {
            String[] arr = e.split(" ");
            char cmd = arr[0].charAt(0);
            int num = Integer.parseInt(arr[1]);
            
            if (cmd == 'I') {
                q.put(num, q.getOrDefault(num, 0) + 1);
                continue;
            }
            
            if (q.isEmpty())
                continue;
            
            int delNum = num == 1 ? q.lastKey(): q.firstKey();
            if (q.put(delNum, q.get(delNum) - 1) == 1)
                q.remove(delNum);
                
        }
        
        if (q.isEmpty()) {
            return answer;
        }
        answer[0] = q.lastKey();
        answer[1] = q.firstKey();
        
        return answer;
    }
}